From 6a16e2ee93ff8e6cadc5985748b76da2c96b673f Mon Sep 17 00:00:00 2001
From: Aymen Zayet <aymen.zayet@intel.com>
Date: Thu, 8 Nov 2012 17:34:16 +0100
Subject: [PATCH 52/78] wl12xx: Wait for pending thread irq before suspend the
 driver.

BZ: 64231

    10-23 20:42:15.077  3088  3088 I KERNEL  : [ 7709.133621] Suspending console(s) (use no_con
    10-23 20:44:04.288  3088  3088 I KERNEL  : [ 7709.178465] wl1271: suspended
    10-23 20:44:04.288  3088  3088 I KERNEL  : [ 7709.178609] ------------[ cut here ]---------
    10-23 20:44:04.288  3088  3088 I KERNEL  : [ 7709.178676] WARNING: at /data/buildbot/workdi
    10-23 20:44:04.288  3088  3088 I KERNEL  : [ 7709.178700] Hardware name: Medfield
    10-23 20:44:04.288  3088  3088 I KERNEL  : [ 7709.178712] queueing ieee80211 work while goi
    10-23 20:44:04.288  3088  3088 I KERNEL  : [ 7709.178726] Modules linked in: matrix(C) wl12
    10-23 20:44:04.288  3088  3088 I KERNEL  : [ 7709.178826] Pid: 6374, comm: irq/258-wl1271 T
    10-23 20:44:04.288  3088  3088 I KERNEL  : [ 7709.178843] Call Trace:
    10-23 20:44:04.288  3088  3088 I KERNEL  : [ 7709.178889]  [<c1254c37>] warn_slowpath_commo
    10-23 20:44:04.288  3088  3088 I KERNEL  : [ 7709.178945]  [<f8762619>] ? ieee80211_queue_d
    10-23 20:44:04.288  3088  3088 I KERNEL  : [ 7709.178994]  [<f8762619>] ? ieee80211_queue_d
    10-23 20:44:04.288  3088  3088 I KERNEL  : [ 7709.179029]  [<c1254d93>] warn_slowpath_fmt+0
    10-23 20:44:04.288  3088  3088 I KERNEL  : [ 7709.179078]  [<f8762619>] ieee80211_queue_del
    10-23 20:44:04.288  3088  3088 I KERNEL  : [ 7709.179125]  [<f855f661>] wl1271_ps_elp_sleep
    10-23 20:44:04.288  3088  3088 I KERNEL  : [ 7709.179164]  [<f855765f>] wl1271_irq+0x1cf/0x
    10-23 20:44:04.288  3088  3088 I KERNEL  : [ 7709.179210]  [<c12ad133>] irq_thread_fn+0x23/
    10-23 20:44:04.288  3088  3088 I KERNEL  : [ 7709.179242]  [<c12ad08c>] irq_thread+0x13c/0x
    10-23 20:44:04.288  3088  3088 I KERNEL  : [ 7709.179272]  [<c12ad110>] ? irq_thread+0x1c0/
    10-23 20:44:04.288  3088  3088 I KERNEL  : [ 7709.179306]  [<c12acf50>] ? irq_finalize_ones
    10-23 20:44:04.288  3088  3088 I KERNEL  : [ 7709.179336]  [<c1279924>] kthread+0x74/0x80
    10-23 20:44:04.288  3088  3088 I KERNEL  : [ 7709.179369]  [<c12798b0>] ? __init_kthread_wo
    10-23 20:44:04.288  3088  3088 I KERNEL  : [ 7709.179406]  [<c19012ba>] kernel_thread_helpe
    10-23 20:44:04.288  3088  3088 I KERNEL  : [ 7709.179427] ---[ end trace 2e4e4cd2bb4f24ee ]

    The driver is already suspended when the thread irq tries to queue a new
    elp_work to ieee80211 workqueue. Since the mac80211 is already suspended,
    the work will not be queued, as a consequence , the device remains powered
    ON during suspend.

    The suspend should wait for the pending thread irq "wl1271_irq()". A synchronize_irq()
    before flushing elp_work should fix this issue.

Change-Id: I8c61c88d156c8611cabf4190fe167bb93e6c0918
Orig-Change-Id: faa41008e5e58a6560adc56f4b2fc7cde2a10adb
Signed-off-by: Aymen Zayet <aymen.zayet@intel.com>
Reviewed-on: http://android.intel.com:8080/73257
Reviewed-by: lab_cactus <lab_cactus@intel.com>
Reviewed-by: De Cesco, Jonathan <jonathan.de.cesco@intel.com>
Tested-by: Champciaux, Nicolas <nicolas.champciaux@intel.com>
Reviewed-by: buildbot <buildbot@intel.com>
Tested-by: buildbot <buildbot@intel.com>
Reviewed-on: http://android.intel.com:8080/78009
Reviewed-by: Li, XiaojingX <xiaojingx.li@intel.com>
Tested-by: Li, XiaojingX <xiaojingx.li@intel.com>
---
 wl12xx-compat/drivers/net/wireless/wl12xx/main.c |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

diff --git a/wl12xx-compat/drivers/net/wireless/wl12xx/main.c b/wl12xx-compat/drivers/net/wireless/wl12xx/main.c
index 51dbcae..7ba4882 100644
--- a/wl12xx-compat/drivers/net/wireless/wl12xx/main.c
+++ b/wl12xx-compat/drivers/net/wireless/wl12xx/main.c
@@ -2301,6 +2301,7 @@ static int wl1271_op_suspend(struct ieee80211_hw *hw,
 
 	wl1271_enable_interrupts(wl);
 	flush_work(&wl->tx_work);
+	synchronize_irq(wl->irq);
 	flush_delayed_work(&wl->elp_work);
 
 	return 0;
-- 
1.7.9

